src/udev/cryptsetup_id/cryptsetup_id.c

```diff
@@ -0,0 +1,147 @@
+#include "udev-util.h"
+
+static const char *arg_device = NULL;
+
+static int help(void) {
+        printf("%s [OPTIONS...] [PATH]\n\n"
+               "  -h --help     Show this help text\n"
+               "  -v --version  Show package version\n",
+               program_invocation_short_name);
+        return 0;
+}
+
+static int parse_argv(int argc, char *argv[]) {
+        static const struct option options[] = {
+                { "help",     no_argument, NULL, 'h' },
+                { "version",  no_argument, NULL, 'v' },
+                {}
+        };
+        int c;
+
+        while ((c = getopt_long(argc, argv, "hv", options, NULL)) >= 0)
+                switch (c) {
+                case 'h':
+                        return help();
+                case 'v':
+                        return version();
+                case '?':
+                        return -EINVAL;
+                default:
+                        assert_not_reached();
+                }
+
+        if (argc > 2)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Error: unexpected argument.");
+
+        arg_device = argv[optind];
+        return 1;
+}
+
+static int print_attribute(const char *name, const char *devname) {
+        _cleanup_(sd_device_unrefp) struct sd_device *dev = NULL;
+        const char *devpath;
+        int r;
+
+        r = sd_device_new_from_devname(&dev, devname);
+        if (r < 0)
+                return log_debug_errno(r, "Could not resolve device '%s': %m", devname);
+
+        if (!device_in_subsystem(dev, "block"))
+                return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Device %s is not a block device", devname);
+
+        r = sd_device_get_devpath(dev, &devpath);
+        if (r < 0)
+                return log_debug_errno(r, "Could not resolve devpath for device '%s': %m", devname);
+
+        printf("CRYPT_%s=%s\n", name, devpath);
+        return 0;
+}
+
+static int run(int argc, char *argv[]) {
+        _cleanup_(sd_device_unrefp) struct sd_device *dev = NULL;
+        _cleanup_(crypt_freep) struct crypt_device *cd = NULL;
+        const char *dm_name, *type;
+        int r;
+
+        (void) udev_parse_config();
+        log_setup();
+
+        r = parse_argv(argc, argv);
+        if (r <= 0)
+                return r;
+
+        if (arg_device) {
+                r = sd_device_new_from_path(&dev, arg_device);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to get device from path %s: %m", arg_device);
+        } else {
+                r = device_new_from_strv(&dev, environ);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to get current device from environment: %m");
+        }
+
+        r = sd_device_get_sysattr_value(dev, "dm/name", &dm_name);
+        if (r < 0)
+                return log_device_debug_errno(dev, r, "Failed to get device DM name: %m");
+
+        r = crypt_init_by_name(&cd, dm_name);
```

Comment:
hmm, does this actually work as it should? i mean, i know that libcryptsetup for some of its ops synchronizes on udev, and if we call these functions from udev's own synchronous workers, then we might deadlock here? I think at the very least there's some flag you might need to set, that ensures the udev synchronization is turned off.